lib/corefoundation/base.rb

```diff
@@ -1,101 +1,34 @@
-    typedef :ulong, :cfoptionflags
-    typedef :ulong, :cftypeid
-    typedef :ulong, :cfhashcode
-  end
-
-
-  # @private
-  class Range < FFI::Struct
-    layout :location, :cfindex,
-           :length, :cfindex
-  end
-
-  typedef :pointer, :cftyperef
-
-  #general utility functions
-  attach_function :show, 'CFShow', [:cftyperef], :void
-  attach_function :release, 'CFRelease', [:cftyperef], :void
-  attach_function :retain, 'CFRetain', [:cftyperef], :cftyperef
-  attach_function 'CFEqual', [:cftyperef, :cftyperef], :char
-  attach_function 'CFHash', [:cftyperef], :cfhashcode
-  attach_function 'CFCopyDescription', [:cftyperef], :cftyperef
-  attach_function 'CFGetTypeID', [:cftyperef], :cftypeid
-
-  # The base class for all of the wrapper classes 
-  #
-  # @abstract
-  class Base  
-    @@type_map = {}
-
-    # @private
-    class Releaser
-      def initialize(ptr)
-        @address  = ptr.address
-      end
-
-      def call *ignored
-        if @address != 0
-          CF.release(@address)
-          @address = 0
-        end
-      end
+    # Raises an exception if the argument does not inherit from CF::Base
+    #
+    # @param cftyperef object to check
+    def self.check_cftype(cftyperef)
+      raise TypeError, "#{cftyperef.inspect} is not a cftype" unless cftyperef.is_a?(CF::Base)
     end
 
-    class << self
-      # Raises an exception if the argument does not inherit from CF::Base
-      #
-      # @param cftyperef object to check
-      def check_cftype(cftyperef)
-        raise TypeError, "#{cftyperef.inspect} is not a cftype" unless cftyperef.is_a?(CF::Base)
-      end
-
-      # Wraps an FFI::Pointer with the appropriate subclass of CF::Base
-      # If the pointer is not a CFTypeRef behaviour is undefined
-      # @param [FFI::Pointer] cftyperef Object to wrap
-      # @return A wrapper object inheriting from CF::Base
-      def typecast(cftyperef)
-        klass = klass_from_cf_type cftyperef
-        klass.new(cftyperef)
-      end
-
-      private
-      def klass_from_cf_type cftyperef
-        klass = @@type_map[CF.CFGetTypeID(cftyperef)]
-        if !klass
-          raise TypeError, "No class registered for cf type #{cftyperef.inspect}"
-        end
-        klass
-      end
-
-      def register_type(type_name)
-        CF.attach_function "#{type_name}GetTypeID", [], :cftypeid
-        @@type_map[CF.send("#{type_name}GetTypeID")] = self
-      end
-
+    # Wraps an FFI::Pointer with the appropriate subclass of CF::Base
+    # If the pointer is not a CFTypeRef behaviour is undefined
+    #
+    # @param [FFI::Pointer] cftyperef Object to wrap
+    # @return A wrapper object inheriting from CF::Base
+    def self.typecast(cftyperef)
+      klass = klass_from_cf_type cftyperef
+      klass.new(cftyperef)
     end
 
-    # 
-    #
-    # @param [FFI::Pointer] ptr The pointer to wrap
-    def initialize(ptr)
-      @ptr = FFI::Pointer.new(ptr)
+    # @param [FFI::Pointer] pointer The pointer to wrap
```

Comment:
I think this is a breaking change - existing documentation states that calling Base.typecast or .new on a pointer doesn't do anything memory management related to the pointer

automating this seems like a positive change though.

I think this leaks though - the proc created here retains self, so the finalizer can never run. For example, if I run

```require 'ffi'
class Type1
  attr_reader :ptr
  def initialize
    @ptr = FFI::Pointer.new 0
    ObjectSpace.define_finalizer(ptr, proc {puts 'type 1 finalizer running'})
  end
end

class Type2
  attr_reader :ptr
  def initialize
    @ptr = FFI::Pointer.new 0
  end
end

10.times do
  ObjectSpace.define_finalizer(Type2.new.ptr, proc {puts 'type 2 finalizer running'})

  Type1.new
end

GC.start
```

then i get the output

```
type 2 finalizer running
type 2 finalizer running
type 2 finalizer running
type 2 finalizer running
type 2 finalizer running
type 2 finalizer running
type 2 finalizer running
type 2 finalizer running
type 2 finalizer running
type 2 finalizer running
```

See https://www.mikeperham.com/2010/02/24/the-trouble-with-ruby-finalizers/ for another approach involving a class method rather than the separate Releaser class

Reply:
Thanks for pointing this out! Pushing an update soon.

Reply:
I've updated it in https://github.com/chef/corefoundation/pull/10/commits/a6f63cf5e47a424b7fe0ca99033c958d75094980 and checked it on local. Seems to be working as expected now. 👌🏼